### PR TITLE
Make it possible to inject Keycloak admin client

### LINKED ISSUE #26837: Support `Keycloak` injection in `quarkus-keycloak-admin-client` and `quarkus-keycloak-admin-client-reactive`
### Description

Right now, the users how to initialize `Keycloak` directly in the code, using a typical sequence:
```
KeycloakBuilder.builder()
                .serverUrl(KEYCLOAK_SERVER_URL)
                .realm("master")
                .clientId("admin-cli")
                .username("admin")
                .password("admin")
                .build();
```

However, as shown by @omar-zahid  in #26836, it would really be simpler to have an injected `Keycloak`, which would also provide some extra value to using `quarkus-keycloak-admin-client`.

The same applies to `quarkus-keycloak-admin-client-reactive`

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
closes: #26837

Keycloak admin client can be configured in `application.properties` and injected as a CDI bean.

### CODE REVIEW

**Comment by michalvavrik** [comment]
cc @sberyozkin

**Comment by michalvavrik** [comment]
Tests seems related, but are passing locally for me, I'll have a look what's different.

File: bom/application/pom.xml
Comment (gsmet) [inline bom/application/pom.xml:886]: What were your reasons to make it a separate extension? CDI is usually at the core of all our extensions and we don't have a separate one to bring CDI support.

Note that there might be good reasons but we need to carefully discuss this before thinking about merging this.

**Review by gsmet — DISMISSED** [review]

File: bom/application/pom.xml
Comment (michalvavrik) [reply bom/application/pom.xml:886]: @gsmet there is classic and reactive Keycloak admin client and we want (I did :-)) to use one CDI on both of them.

Comment (michalvavrik) [reply bom/application/pom.xml:886]: Please note that I added unlisted flag to true, so in my understanding the CDI won't be listed to users...

Comment (sberyozkin) [reply bom/application/pom.xml:886]: Hi @michalvavrik Thanks for the PR, I'll have a deeper look next week, but I did not expect that 2 new modules would have to be added too :-), I'd just have it all bundled in with the existing extension, there is not much there to add, the config group, and provding a synthetic bean, which s what we do for OIDC, and everywhere else when the injection support is needed for non real beans. Think about it please, thanks

Comment (michalvavrik) [reply bom/application/pom.xml:886]: Alright, as I've time before you review this, I'll try to think about it. 

Just to be honest, I'm not sure if you guys thought about it (might be wrong!), because following classes would have to be duplicated (f.e. you can't bring `quarkus-core-deployment` to runtime deps and so on): `io.quarkus.keycloak.adminclient.cdi.deployment.KeycloakAdminClientCDIProcessor`, `io.quarkus.keycloak.adminclient.cdi.runtime.KeycloakAdminClientRecorder`, and build items....

I could extract business logic to an abstraction level (one new module) and duplicate build steps and recorder that would merely call the abstraction level....

Comment (gsmet) [reply bom/application/pom.xml:886]: That would be better than having user-facing artifacts. Better hide things from the user. If you have a look at some other extensions, we are using this very pattern (with some `-common` extensions).

The important point is that CDI is not optional in Quarkus: it's at its very core. Thus our extensions are CDI-enabled by default.

**Comment by michalvavrik** [comment]
FYI - @sberyozkin I reflected both of your comments, but I'll push the changes tomorrow afternoon as I need re-test it after refactoring. Thank you for the feedback.

File: bom/application/pom.xml
Comment (michalvavrik) [reply bom/application/pom.xml]: CDI is now integral part of keycloak-admin-client/reactive, the new modules can't be avoided:
- `common` for config and validation
- `SPI` for admin client <-> oidc

I checked existing extensions several times and can't see significant the difference in current approach, but I'm happy to learn/adjust if you do! Thank you

**Comment by michalvavrik** [comment]
I checked test failures - not related.

**Comment by sberyozkin** [comment]
@michalvavrik Hi, I'm sorry, but I'm still not understanding the need for the SPI module as well a few other things in the PR, why can't we just have a synthetic bean produced in the recorder, which will do:
```
KeycloakBuilder.builder()
                .serverUrl(KEYCLOAK_SERVER_URL)
                .realm("master")
                .clientId("admin-cli")
                .username("admin")
                .password("admin")
                .build();
```

where `serverUrl`, `realm`, `clientId`, `username` and `password` are configured ?

File: docs/src/main/asciidoc/security-keycloak-admin-client.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-keycloak-admin-client.adoc:119]: I'm not sure we need this kind of flexibility to support multiple named KeycloakAdmins, if we ever do then we could do similarly to the way OIDC tenants or OIDC clients are supported but for now I think we should have:
```
quarkus.keycloak.admin-client.client-id=...
```
etc.
Note `client-name` is ok as a property but we already have `client-id` in the OIDC/OIDC Client config...

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java
Comment (sberyozkin) [inline extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java:20]: I'm a bit confused what is happening here. Well, I think you have some interesting ideas here, but updating OIDC Dev Services should be, IMHO, a separate issue with a dedicated discussion. IMHO this PR will become much simpler :-) if we focus here on supporting a basic `Keycloak` injection, and then we can continue from there

File: docs/src/main/asciidoc/security-keycloak-admin-client.adoc
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-admin-client.adoc:119]: How do you know users in majority of cases want to create exactly one client? There is no extra cost to creating to. `client-name` is the name of the client (named bean).

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java
Comment (michalvavrik) [reply extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java:20]: Well, this simply adds configuration option for client roles and default client scope as explained in the docs. How could you manage Keycloak with `microprofile-jwt` and without possibility to add users created by dev services new client roles (view users, view roles).

I can remove it, but once again, whole point of Dev Services is to allow user hit the ground running. With single line (as clearly showed in UsersConfiguratios test I added) allows user to configure. Doing same with keycloak.json takes much longer.

Comment (michalvavrik) [reply extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java:20]: And it's more complicated, not every developer is Keycloak expert...

**Comment by michalvavrik** [comment]
Thank you for feedback as always. I will need confirmation that you mean your comments and I should drop it all.

1. It makes no sense to me and plenty (plenty) of work went into this, but I respect your opinion and will do whatever you want. It's all showed in tests.
2. I just want to be sure you really read the PR before I drop it, that's all...

Just write `I am sure`.

**Comment by michalvavrik** [comment]
Also, how else would you configure Keycloak admin client with Keycloak Dev Services without SPI?

File: docs/src/main/asciidoc/security-keycloak-admin-client.adoc
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-admin-client.adoc:119]: In https://github.com/quarkusio/quarkus/issues/26837#issuecomment-1190837906 you wrote `I believe we'll need to have a KeycloakAdminClient group.` which I though suggest multiple clients too.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-admin-client.adoc:119]: Please allow me to also provide one else point of view - allowing to configure one client looks good in "quick start"/docs, but in production you will hardly ever have one almighty client with all the powers. You are more likely to give access on need to have basis, which suggest you need more clients for more scenarios. It make sense for users to use our configuration instead of creating the beans themselves as otherwise they need to add some hook to close the client themselves.

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java
Comment (sberyozkin) [reply extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakBuildTimeConfig.java:20]: @michalvavrik Sure, I'd just prefer this enhancement it done in a follow up PR as this one is only about  supporting an injection as opposed to a manual construction of Keycloak

**Comment by sberyozkin** [comment]
@michalvavrik Let me take a quick stab asap at what I meant, as it feels like we might be talking about different things :-)
Of course, your effort is appreciated, I'm sorry if it well end up that some of your code may need to be changed - but it has happened to me a lot, including having to completely rewriting an original DevServices for KC module structure :-)

**Comment by michalvavrik** [comment]
I appreciate the response. I understand my ideas were bad and I'm happy to drop them (even though I'd really appreciate your comment/response on `authorization` as these thoughts I had came from my previous practice were we used Keycloak, after all, you just suggest to ignore `KeycloakBuilder#authorization` method... I'll drop it anyway!).

As I'm OOTO and very busy, I'll move the PR to draft and rewrite it 29.8.. Thank you for your time.

**Comment by gsmet** [comment]
@michalvavrik @sberyozkin maybe you both could have a talk so that you agree on a common plan?

**Comment by sberyozkin** [comment]
@michalvavrik 

>I understand my ideas were bad

No, I don't think I even implied it, and I do appreciate you were trying to innovate here :-). But IMHO it makes it much simpler to handle it step by step, right now, I believe it looks much closer to the original goal of #26837. We have a much simpler way to deal with initializing `Keycloak` admin object. 

I'm not sure about the config of the grants though yet, I'll look asap.

Re some of your earlier points:
> Keycloak admin client with Keycloak Dev Services without SPI

I'm not sure why we need to do it. `Keycloak Dev Services` work right now without Keycloak admin client. Does it have to use it ? May be but lets open an enhancement request and discuss the reason why we need to bring this dependency, I'm not sure we need to but may be I'm missing something.

> Please allow me to also provide one else point of view - allowing to configure one client looks good in "quick start"/docs, but in production you will hardly ever have one almighty client with all the powers.

Well, it makes sense, in general. We have `OidcClient` able to support many targets. Do we need the same for `Keycloak` admin client, given that a single one can setup everything withing a given Keycloak instance, I'm not sure, but if we do need, than may be we should do it similarly to the multi-tenant setup of the oidc.

All of this can be useful for Quarkus users, but my point is, lets handle it step by step, with the dedicated issues

Thanks for your help so far

**Comment by sberyozkin** [comment]
@michalvavrik FYI: https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientsConfig.java, and there is a support for injecting named clients in the deployment module.
Have a look at in the next step please.

File: extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java
Comment (sberyozkin) [inline extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java:61]: @michalvavrik I believe it has to be defaulted to `password` since the typical use is captured in this removed code:
```
    private Keycloak keycloak() {
        try {
            return KeycloakBuilder.builder()
                    .serverUrl(url)
                    .realm("master")
                    .clientId("admin-cli")
                    .username("admin")
                    .password("admin")
                    .build();
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
```

where a `password` grant is used to get access to the `master` realm - from where other realms can be created.

File: extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java
Comment (sberyozkin) [inline extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java:30]: I propose `admin-cli` by default, we have a default `master` realm configured here too

File: extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java
Comment (sberyozkin) [inline extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java:37]: I'll leave it up to you, but since `admin:admin` is used a lot in test/demo instances of KC, incl the one launched by DevServices for KC, I'd have no problems with defaulting the username and password to `admin`.

So in a typical case where a user wants to use it to setup a test KC instance, one would only have to set the `server-url`

Comment (michalvavrik) [reply extensions/keycloak-admin-client-common/runtime/src/main/java/io/quarkus/keycloak/admin/client/common/KeycloakAdminClientConfig.java:43]: I tried to avoid "necessity" to remove password/username with client_credentials:
```
quarkus.keycloak.admin-client.username=
quarkus.keycloak.admin-client.password=
```
though implementation don't set them with client_credentials anyway https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/integration/admin-client/src/main/java/org/keycloak/admin/client/token/TokenManager.java#L79, it feels wrong to rely on impl.

It looks better with defaults, so I've added them. Thank you

**Comment by michalvavrik** [comment]
Sounds good, I can see three issues (topics): authorization, dev services, multiple clients. Once this PR is done, I'll create an issue for Dev Services for Keycloak where it can be discussed whether it makes sense. If so, I'll provide impl. Thank you

File: docs/src/main/asciidoc/security-keycloak-admin-client.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-keycloak-admin-client.adoc:118]: Hi @michalvavrik It all looks good to me now, thanks, one last thing I'd like to suggest. I believe, to avoid the confusion (`quarkus-oidc-client` is designed to request the tokens in a provider API neutral way), it makes sense to add a note somewhere under this example, which is showing how to get a token using a client cred grant, to add a note like "Note that the oidc-client can also be used to acquire tokens" and link to the guide, something like that

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-admin-client.adoc:118]: Sure, done. I know you can use the Keycloak admin client to get a token, but I thought (expected) the other way around (you can use OIDC client to get a token that you set when you are creating the admin client). Thank you.

**Review by sberyozkin — APPROVED** [review]

**Pull request merged**
